cmd/server/main.go

```diff
@@ -187,96 +190,82 @@ func bundle(w http.ResponseWriter, r *http.Request) {
-			AuthFilePath:        "/dev/null",
-			DockerAuthConfig: &types.DockerAuthConfig{
-				Username:      "",
-				Password:      "",
-				IdentityToken: "",
-			},
-		}
-		_, err = copy.Image(r.Context(), policyContext, localRef, srcRef, &copy.Options{
-			RemoveSignatures:      true,
-			SignBy:                "",
-			ForceManifestMIMEType: "",
-			DestinationCtx:        destCtx,
-			SourceCtx:             srcCtx,
-		})
-		if err != nil {
-			err = errors.Wrapf(err, "failed to save docker image archive of %s", image)
-			handleHttpError(w, r, err, http.StatusInternalServerError)
-			return
-		}
-
-		f, err := os.Open(destPath)
-		if err != nil {
-			err = errors.Wrap(err, "failed to open override image archive")
-			handleHttpError(w, r, err, http.StatusInternalServerError)
-			return
-		}
-		fi, err := f.Stat()
-		if err != nil {
-			handleHttpError(w, r, err, http.StatusInternalServerError)
-			return
-		}
-
-		header := &tar.Header{
-			Name:    fmt.Sprintf("kurl/image-overrides/%d.tar", i),
-			Size:    fi.Size(),
-			Mode:    0644,
-			ModTime: time.Now(),
+	errCh := make(chan error, 1)
+	defer close(errCh)
+
+	imageCh := make(chan string, len(bundle.Images))
+
+	// HACK: Work around the fact that github.com/containers/image/v5/copy copy.Image does not support streaming.
+	// Download images concurrently while streaming the rest of the bundle to minimize idle time.
+	go func() {
+		defer close(imageCh)
+
+		for i, image := range bundle.Images {
+			srcRef, err := alltransports.ParseImageName(fmt.Sprintf("docker://%s", image))
+			if err != nil {
+				errCh <- errors.Wrapf(err, "error parsing override image %q: %v", image, err)
+				return
+			}
+
+			destPath := path.Join(tempDir, fmt.Sprintf("%d.tar", i))
+			destStr := fmt.Sprintf("docker-archive:%s:%s", destPath, image)
+			localRef, err := alltransports.ParseImageName(destStr)
+			if err != nil {
+				errCh <- errors.Wrapf(err, "failed to parse local image name: %s", destStr)
+				return
+			}
+
+			policy, err := signature.NewPolicyFromBytes(imagePolicy)
+			if err != nil {
+				errCh <- errors.Wrapf(err, "failed to read default image policy for image %q", image)
+				return
+			}
+			policyContext, err := signature.NewPolicyContext(policy)
+			if err != nil {
+				errCh <- errors.Wrapf(err, "failed to create image policy context for image %q", image)
+				return
+			}
+			destCtx := &types.SystemContext{
+				DockerDisableV1Ping: true,
+			}
+			srcCtx := &types.SystemContext{
+				DockerDisableV1Ping: true,
+				AuthFilePath:        "/dev/null",
+				DockerAuthConfig: &types.DockerAuthConfig{
+					Username:      "",
+					Password:      "",
+					IdentityToken: "",
+				},
+			}
+			_, err = copy.Image(r.Context(), policyContext, localRef, srcRef, &copy.Options{
+				RemoveSignatures:      true,
+				SignBy:                "",
+				ForceManifestMIMEType: "",
+				DestinationCtx:        destCtx,
+				SourceCtx:             srcCtx,
+			})
+
+			runtime.GC()
```

Comment:
am not 100% sure, since[ runtime.GC() ](https://pkg.go.dev/runtime#GC)is going to block, can it be run outside the for loop only once?

Reply:
@divolgin can you help me understand why this is necessary? this code was lifted from kots i believe.

Reply:
to keep memory usage down.  because Go will keep requesting more memory from the system for each image pull, eventually the pod will be evicted

Reply:
so sounds like this should stay in the loop. thanks